File: Prefix & Postfix_Suffix Sum/Subarray with Sum Zero/subarray_with_sum_zero.cpp
#include<bits/stdc++.h>
using namespace std;

// O(N^2) Time Complexity
/* bool subarrayWithSumZero(vector<int>& arr){
    int n=arr.size();
    for(int i=0;i<n;i++){
        int sum=arr[i];
        if(sum==0){
            return true;
        }
        for(int j=i+1;j<n;j++){
            sum+=arr[j];
            if(sum==0){
                return true;
            }
        }
    }
    return false;
} */

bool subarrayWithSumZero(vector<int>& arr){
    unordered_set<int> sumSet;
    int sum=0;
    for(int i=0;i<arr.size();i++){
        sum+=arr[i];
        if(sum==0 || sumSet.find(sum)!=sumSet.end()){
            return true;
        }else{
            sumSet.insert(sum);
        }
    }
    return false;
}

int main(){
    // vector<int> arr={3,-2,1,1,4};
    vector<int> arr={3,2,1,1,4};
    if(subarrayWithSumZero(arr)){
        cout<<"True"<<endl;
    }else{
        cout<<"False"<<endl;
    }
    return 0;
}

File: Prefix & Postfix_Suffix Sum/Subarray with Sum Zero/subarray_with_sum_zero.py
class Solution:
    def subArrayExists(self, arr):
        sumSet = set()
        sum = 0
        for i in arr:
            sum += i
            if sum == 0 or sum in sumSet:
                return True
            else:
                sumSet.add(sum)
        return False


if __name__ == "__main__":
    obj = Solution()
    arr = [4, 2, -3, 1, 6]
    print(obj.subArrayExists(arr=arr))
